File: android/app/src/main/java/com/rork/calzyandroid/ui/sheets/EditMealSheet.kt
package com.rork.calzyandroid.ui.sheets

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Add
import androidx.compose.material.icons.outlined.Close
import androidx.compose.material.icons.outlined.Remove
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.rork.calzyandroid.AppViewModel
import com.rork.calzyandroid.data.MealEntry
import com.rork.calzyandroid.data.Nutrition
import com.rork.calzyandroid.ui.components.CalzyCard
import com.rork.calzyandroid.ui.components.FullScreenSheet
import com.rork.calzyandroid.ui.components.MetricText
import com.rork.calzyandroid.ui.components.Pressable
import com.rork.calzyandroid.ui.components.PrimaryButton
import com.rork.calzyandroid.ui.theme.CalzyColors
import com.rork.calzyandroid.ui.theme.MetricFontFamily

private const val MAX_CALORIES = 20000

/**
 * Quick-correction form for an already-logged meal: rename it or fix its
 * calories. Macros scale with the calorie edit so the split stays believable.
 */
@Composable
fun EditMealSheet(
    meal: MealEntry?,
    viewModel: AppViewModel,
    onClose: () -> Unit,
    onSaved: (MealEntry) -> Unit,
) {
    var title by remember { mutableStateOf("") }
    var calories by remember { mutableIntStateOf(0) }

    LaunchedEffect(meal?.id) {
        if (meal != null) {
            title = meal.title
            calories = Nutrition.mealCalories(meal)
        }
    }

    val originalCalories = meal?.let { Nutrition.mealCalories(it) } ?: 0
    val preview = meal?.let { Nutrition.mealWithCalories(it, calories) }
    val trimmed = title.trim()
    val canSave = meal != null && trimmed.isNotEmpty() &&
        (trimmed != meal.title || calories != originalCalories)
    val delta = calories - originalCalories

    fun step(amount: Int) {
        calories = (calories + amount).coerceIn(0, MAX_CALORIES)
    }

    FullScreenSheet(
        open = meal != null,
        onClose = onClose,
        title = "Edit meal",
        footer = {
            PrimaryButton(
                text = "Save changes",
                enabled = canSave,
                onClick = {
                    if (meal == null || !canSave) return@PrimaryButton
                    val updated = Nutrition.mealWithCalories(meal, calories).copy(title = trimmed)
                    viewModel.updateMeal(updated)
                    onSaved(updated)
                    onClose()
                },
            )
        },
    ) {
        if (meal == null || preview == null) return@FullScreenSheet
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 20.dp)
                .padding(top = 6.dp, bottom = 32.dp),
            verticalArrangement = Arrangement.spacedBy(14.dp),
        ) {
            // Name
            CalzyCard(radius = 22.dp) {
                Column(modifier = Modifier.padding(16.dp)) {
                    Text(
                        text = "Meal name",
                        fontSize = 13.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = CalzyColors.inkSoft,
                        modifier = Modifier.padding(bottom = 9.dp),
                    )
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        modifier = Modifier
                            .fillMaxWidth()
                            .clip(RoundedCornerShape(16.dp))
                            .background(CalzyColors.well),
                    ) {
                        TextField(
                            value = title,
                            onValueChange = { title = it },
                            placeholder = {
                                Text(
                                    text = "e.g. Chicken burrito bowl",
                                    color = CalzyColors.inkFaint,
                                )
                            },
                            textStyle = TextStyle(
                                fontSize = 17.sp,
                                fontWeight = FontWeight.SemiBold,
                                color = CalzyColors.ink,
                            ),
                            modifier = Modifier.weight(1f),
                            colors = TextFieldDefaults.colors(
                                focusedContainerColor = Color.Transparent,
                                unfocusedContainerColor = Color.Transparent,
                                focusedIndicatorColor = Color.Transparent,
                                unfocusedIndicatorColor = Color.Transparent,
                            ),
                            singleLine = true,
                        )
                        if (title.isNotEmpty()) {
                            Pressable(onClick = { title = "" }) {
                                Icon(
                                    imageVector = Icons.Outlined.Close,
                                    contentDescription = "Clear meal name",
                                    tint = CalzyColors.inkFaint,
                                    modifier = Modifier
                                        .padding(end = 14.dp)
                                        .size(16.dp),
                                )
                            }
                        }
                    }
                }
            }

            // Calories
            CalzyCard(radius = 22.dp) {
                Column(modifier = Modifier.padding(16.dp)) {
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Text(
                            text = "Calories",
                            fontSize = 13.sp,
                            fontWeight = FontWeight.SemiBold,
                            color = CalzyColors.inkSoft,
                            modifier = Modifier.weight(1f),
                        )
                        if (delta != 0) {
                            val deltaTint = if (delta > 0) CalzyColors.flame else CalzyColors.mint
                            Box(
                                modifier = Modifier
                                    .clip(CircleShape)
                                    .background(deltaTint.copy(alpha = 0.12f))
                                    .padding(horizontal = 8.dp, vertical = 4.dp),
                            ) {
                                Text(
                                    text = "${if (delta > 0) "+$delta" else "$delta"} kcal",
                                    fontSize = 12.sp,
                                    fontWeight = FontWeight.SemiBold,
                                    color = deltaTint,
                                )
                            }
                        }
                    }

                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(top = 14.dp),
                        horizontalArrangement = Arrangement.spacedBy(18.dp, Alignment.CenterHorizontally),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        StepCircle(
                            label = "Decrease calories",
                            icon = Icons.Outlined.Remove,
                            enabled = calories > 0,
                        ) { step(-10) }

                        Column(horizontalAlignment = Alignment.CenterHorizontally) {
                            TextField(
                                value = "$calories",
                                onValueChange = { raw ->
                                    val digits = raw.filter { it.isDigit() }
                                    calories = (digits.toIntOrNull() ?: 0).coerceAtMost(MAX_CALORIES)
                                },
                                textStyle = TextStyle(
                                    fontSize = 40.sp,
                                    fontFamily = MetricFontFamily,
                                    fontWeight = FontWeight.ExtraBold,
                                    color = CalzyColors.ink,
                                    textAlign = TextAlign.Center,
                                ),
                                modifier = Modifier.width(150.dp),
                                colors = TextFieldDefaults.colors(
                                    focusedContainerColor = Color.Transparent,
                                    unfocusedContainerColor = Color.Transparent,
                                    focusedIndicatorColor = Color.Transparent,
                                    unfocusedIndicatorColor = Color.Transparent,
                                ),
                                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                                singleLine = true,
                            )
                            Text(
                                text = "kcal",
                                fontSize = 12.sp,
                                fontWeight = FontWeight.Medium,
                                color = CalzyColors.inkFaint,
                            )
                        }

                        StepCircle(
                            label = "Increase calories",
                            icon = Icons.Outlined.Add,
                            enabled = calories < MAX_CALORIES,
                        ) { step(10) }
                    }

                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(top = 14.dp),
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                    ) {
                        listOf(-100, -50, 50, 100).forEach { amount ->
                            Pressable(onClick = { step(amount) }, modifier = Modifier.weight(1f)) {
                                Box(
                                    modifier = Modifier
                                        .fillMaxWidth()
                                        .clip(CircleShape)
                                        .background(CalzyColors.well)
                                        .padding(vertical = 9.dp),
                                    contentAlignment = Alignment.Center,
                                ) {
                                    Text(
                                        text = if (amount > 0) "+$amount" else "$amount",
                                        fontSize = 13.sp,
                                        fontWeight = FontWeight.SemiBold,
                                        color = CalzyColors.ink,
                                    )
                                }
                            }
                        }
                    }
                }
            }

            // Macro preview
            CalzyCard(radius = 22.dp) {
                Column(modifier = Modifier.padding(16.dp)) {
                    Text(
                        text = "Macros scale to match",
                        fontSize = 13.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = CalzyColors.inkSoft,
                    )
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(top = 12.dp),
                        horizontalArrangement = Arrangement.spacedBy(10.dp),
                    ) {
                        MacroPill(
                            name = "Protein",
                            value = Nutrition.mealProtein(preview),
                            tint = CalzyColors.protein,
                            modifier = Modifier.weight(1f),
                        )
                        MacroPill(
                            name = "Carbs",
                            value = Nutrition.mealCarbs(preview),
                            tint = CalzyColors.carbs,
                            modifier = Modifier.weight(1f),
                        )
                        MacroPill(
                            name = "Fat",
                            value = Nutrition.mealFat(preview),
                            tint = CalzyColors.fat,
                            modifier = Modifier.weight(1f),
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun StepCircle(
    label: String,
    icon: androidx.compose.ui.graphics.vector.ImageVector,
    enabled: Boolean,
    onClick: () -> Unit,
) {
    Pressable(onClick = onClick, enabled = enabled) {
        Box(
            modifier = Modifier
                .size(40.dp)
                .clip(CircleShape)
                .background(CalzyColors.well),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                imageVector = icon,
                contentDescription = label,
                tint = if (enabled) CalzyColors.ink else CalzyColors.inkFaint,
                modifier = Modifier.size(15.dp),
            )
        }
    }
}

File: ios-calzy/ModernBodyFoodScanner/Views/EditMealSheet.swift
import SwiftUI

/// Quick-correction form for an already-logged meal.
///
/// AI estimates and database matches are close but rarely exact, so this lets a
/// user fix the two things they actually notice on the timeline: what the meal
/// is called and how many calories it cost them. Macros scale with the calorie
/// edit so the split they already saw stays believable.
struct EditMealSheet: View {
    @Environment(AppStore.self) private var store
    @Environment(\.dismiss) private var dismiss

    let meal: MealEntry
    /// Called with the saved meal so a presenting detail view can refresh.
    var onSave: ((MealEntry) -> Void)?

    @State private var title: String
    @State private var calories: Int
    @FocusState private var isNameFocused: Bool

    init(meal: MealEntry, onSave: ((MealEntry) -> Void)? = nil) {
        self.meal = meal
        self.onSave = onSave
        _title = State(initialValue: meal.title)
        _calories = State(initialValue: meal.calories)
    }

    private var trimmedTitle: String {
        title.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private var preview: MealEntry {
        meal.settingCalories(calories)
    }

    private var canSave: Bool {
        !trimmedTitle.isEmpty && (trimmedTitle != meal.title || calories != meal.calories)
    }

    private var delta: Int { calories - meal.calories }

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(spacing: 14) {
                    nameCard
                    caloriesCard
                    macroPreview
                }
                .padding(.horizontal, 20)
                .padding(.top, 6)
                .padding(.bottom, 32)
            }
            .scrollIndicators(.hidden)
            .scrollDismissesKeyboard(.interactively)
            .background(Theme.backdrop)
            .navigationTitle("Edit meal")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarLeading) {
                    Button("Cancel") { dismiss() }
                        .foregroundStyle(Theme.inkSoft)
                }
            }
            .safeAreaInset(edge: .bottom) { saveBar }
        }
        .presentationDetents([.large])
        .presentationDragIndicator(.visible)
    }

    // MARK: - Sections

    private var nameCard: some View {
        VStack(alignment: .leading, spacing: 9) {
            Text("Meal name")
                .font(.system(size: 13, weight: .semibold))
                .foregroundStyle(Theme.inkSoft)

            HStack(spacing: 10) {
                TextField("e.g. Chicken burrito bowl", text: $title)
                    .font(.system(size: 17, weight: .semibold))
                    .foregroundStyle(Theme.ink)
                    .focused($isNameFocused)
                    .submitLabel(.done)
                    .autocorrectionDisabled()

                if !title.isEmpty {
                    Button {
                        Haptics.selection()
                        title = ""
                        isNameFocused = true
                    } label: {
                        Image(systemName: "xmark.circle.fill")
                            .font(.system(size: 16))
                            .foregroundStyle(Theme.inkFaint)
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel("Clear meal name")
                }
            }
            .padding(.horizontal, 14)
            .padding(.vertical, 13)
            .background(Theme.well, in: RoundedRectangle(cornerRadius: 16, style: .continuous))
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .cardStyle(radius: 22, padding: 16)
    }

    private var caloriesCard: some View {
        VStack(spacing: 14) {
            HStack {
                Text("Calories")
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(Theme.inkSoft)
                Spacer()
                if delta != 0 {
                    Text(delta > 0 ? "+\(delta) kcal" : "\(delta) kcal")
                        .font(.system(size: 12, weight: .semibold))
                        .foregroundStyle(delta > 0 ? Theme.flame : Theme.mint)
                        .padding(.horizontal, 8)
                        .padding(.vertical, 4)
                        .background(
                            (delta > 0 ? Theme.flame : Theme.mint).opacity(0.12),
                            in: Capsule()
                        )
                        .transition(.scale.combined(with: .opacity))
                }
            }

            HStack(spacing: 18) {
                stepButton("minus", enabled: calories > 0) { step(-10) }

                VStack(spacing: 2) {
                    TextField("0", value: $calories, format: .number)
                        .keyboardType(.numberPad)
                        .multilineTextAlignment(.center)
                        .font(.metric(40, .bold))
                        .foregroundStyle(Theme.ink)
                        .frame(minWidth: 120)
                    Text("kcal")
                        .font(.system(size: 12, weight: .medium))
                        .foregroundStyle(Theme.inkFaint)
                }

                stepButton("plus", enabled: calories < 20000) { step(10) }
            }

            HStack(spacing: 8) {
                ForEach([-100, -50, 50, 100], id: \.self) { amount in
                    Button {
                        step(amount)
                    } label: {
                        Text(amount > 0 ? "+\(amount)" : "\(amount)")
                            .font(.system(size: 13, weight: .semibold))
                            .foregroundStyle(Theme.ink)
                            .frame(maxWidth: .infinity)
                            .padding(.vertical, 9)
                            .background(Theme.well, in: Capsule())
                    }
                    .buttonStyle(PressableButtonStyle())
                }
            }
        }
        .animation(.spring(response: 0.3, dampingFraction: 0.8), value: delta)
        .frame(maxWidth: .infinity)
        .cardStyle(radius: 22, padding: 16)
    }

    private var macroPreview: some View {
        VStack(spacing: 12) {
            HStack {
                Text("Macros scale to match")
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(Theme.inkSoft)
                Spacer()
            }

            HStack(spacing: 10) {
                macroPill("Protein", preview.protein, Theme.protein)
                macroPill("Carbs", preview.carbs, Theme.carbs)
                macroPill("Fat", preview.fat, Theme.fat)
            }
        }
        .frame(maxWidth: .infinity)
        .cardStyle(radius: 22, padding: 16)
    }

    private var saveBar: some View {
        Button(action: save) {
            HStack(spacing: 8) {
                Image(systemName: "checkmark.circle.fill")
                Text("Save changes")
            }
            .font(.system(size: 17, weight: .semibold))
            .foregroundStyle(.white)
            .frame(maxWidth: .infinity)
            .padding(.vertical, 16)
            .background(canSave ? Theme.ink : Theme.inkFaint, in: Capsule())
        }
        .buttonStyle(PressableButtonStyle())
        .disabled(!canSave)
        .padding(.horizontal, 20)
        .padding(.bottom, 8)
    }

    // MARK: - Pieces

    private func macroPill(_ name: String, _ value: Double, _ color: Color) -> some View {
        VStack(spacing: 4) {
            Text("\(Int(value.rounded()))g")
                .font(.metric(18, .bold))
                .foregroundStyle(Theme.ink)
                .contentTransition(.numericText(value: value))
            Text(name)
                .font(.system(size: 11, weight: .medium))
                .foregroundStyle(Theme.inkSoft)
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, 11)
        .background(color.opacity(0.09), in: RoundedRectangle(cornerRadius: 16, style: .continuous))
    }

    private func stepButton(_ symbol: String, enabled: Bool, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Image(systemName: symbol)
                .font(.system(size: 15, weight: .bold))
                .foregroundStyle(enabled ? Theme.ink : Theme.inkFaint)
                .frame(width: 40, height: 40)
                .background(Theme.well, in: Circle())
        }
        .buttonStyle(PressableButtonStyle())
        .disabled(!enabled)
        .accessibilityLabel(symbol == "plus" ? "Increase calories" : "Decrease calories")
    }

    // MARK: - Actions

    private func step(_ amount: Int) {
        Haptics.selection()
        withAnimation(.spring(response: 0.28, dampingFraction: 0.8)) {
            calories = min(20000, max(0, calories + amount))
        }
    }

    private func save() {
        guard canSave else { return }
        var updated = meal.settingCalories(calories)
        updated.title = trimmedTitle
        store.updateMeal(updated)
        Haptics.success()
        onSave?(updated)
        dismiss()
    }
}
